### ISSUE TITLE
Combined public-private mode

### ISSUE THREAD

**Issue by jetpacktuxedo** [thelounge/thelounge#185]
In some cases it might be useful to run a public server, while still allowing logged in users to take advantage of the private-mode features like persistent logins. 

As far as I can tell, this can only currently be achieved by running two servers simultaneously on different ports, one set to public and the other set to private.

It would be convenient if the public mode still provided a place for users with accounts to log in, without blocking users that do not yet have accounts.

**Comment by jvperrin** [comment]
I agree, I would like to see this feature as well. This is essentially the same request as in erming/shout#251 and would be very useful for providing both a public IRC server while also allowing for users that can sign in to save chat logs and stay connected.

I'd be happy to work on this, but I don't know that much about this project and its structure, so it might take me some time to be able to work on this at all. Great job so far on this though, I only found this project a couple days ago, and it looks very nice. :+1:

**Comment by xPaw** [comment]
@thelounge/maintainers Do we want this or not? My feeling on this is that we shouldn't do this, and update our codebase to correctly check public config variable (I added `isRegistered` at some point that should be removed) and not bind events in public mode at all (e.g. `push:register`).

**Comment by YaManicKill** [comment]
As a theory, I don't have anything against this feature, as it makes sense. But I've not seen anyone want it enough to pick up the code for it. What you say makes sense until this were to happen, though. There's nothing stopping us doing 3 modes, and only binding events that public don't use in private/combined mode. So I don't think that would be a waste.

**Comment by astorije** [comment]
What do we think on this? I don't have anything in favor or against actually, because at the very least one could set up 2 configuration files, one for private and one for public (accessible to different ports/URLs though).
But I can see how it would be useful to land on a login page (for private mode) and see a "Login as guest" option at the bottom or something.

Reason why I'm bringing that up now is the following actually: at the moment, public mode is `public: true` in the config, and private is `public: false`. If that's ever a possibility, or if package system allows for other cases, then maybe we should have `mode: "public"`/`mode: "private"` (to allow for `mode: "both"` or something). That would be for v3.

Alternatively, we could do something slightly different and replace `public` with `authentication: "none"` (i.e. public) / `authentication: "built-in"` (i.e. private) / `authentication: "ldap"` (i.e. `ldap.enable: true`) so that package system could allow alternate authentication mechanisms to be picked from.

So in a nutshell, trying to think ahead of capabilities so we can make the configuration doable in v3 and add features afterwards. Thoughts?

**Comment by xPaw** [comment]
@astorije I was thinking about this recently and came to the conclusion that not doing mixed mode is better over all. The use case of mixing is really niche and limiting. We most likely will end up having more bugs related to this specific feature.

People will most likely also want `lockNetwork` to work for public mode, but allow users to connect to any network and things like that. It's just better served by having 2 separate configs.

**Comment by astorije** [comment]
@xPaw, that's totally fair. I think it wouldn't be too difficult for someone to come up with a recipe that sets up 2 config files for The Lounge, 1 HTML file with the look-and-feel of The Lounge (can re-use CSS) to point out to both, and 1 nginx config that serves this HTML page at `/`, public instance at `/guest`, and private instance at another endpoint. That would be a nice little project.

Will close this issue then.

What about this though?

> we could do something slightly different and replace `public` with `authentication: "none"` (i.e. public) / `authentication: "built-in"` (i.e. private) / `authentication: "ldap"` (i.e. `ldap.enable: true`) so that package system could allow alternate authentication mechanisms to be picked from.

I think the current situation is pretty awkward and limiting. Awkward because we constantly have to check that `if !public && !ldap.enable`. Limiting because if we had other authentication mechanisms or support packages for that (people want to use a MySQL DB to link to their website or whatnot), then we have that issue all over again.
I think we should have one consistent setting for that like described above. Labels TBD obviously.

**Comment by xPaw** [comment]
Closing as discussed.
